maven-embedder/src/main/java/org/apache/maven/cli/props/InterpolationHelper.java

```diff
@@ -0,0 +1,353 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.maven.cli.props;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Function;
+
+public class InterpolationHelper {
+
+    private InterpolationHelper() {}
+
+    private static final char ESCAPE_CHAR = '\\';
+    private static final String DELIM_START = "${";
+    private static final String DELIM_STOP = "}";
+    private static final String MARKER = "$__";
+    private static final String ENV_PREFIX = "env:";
```

Comment:
Why not: `env.` - it is more natural as we can use `${env.NAME}`

Reply:
Good catch. It should be consistent with our current approach.

Reply:
Actually, I'll remove that part, as we do pass the system+"env." properties for substitution already.